tidy3d/components/grid/mesher.py

```diff
@@ -96,66 +96,71 @@ def parse_structures(
+            # indexes of structures that the current structure contains in 2D
+            struct_contains_inds = []
+            for ind, bounds in enumerate(struct_bbox[: struct_ind - 1]):
                 if (
-                    bounds_2d[0] <= plane_bounds[0]
-                    and bounds_2d[1] <= plane_bounds[1]
-                    and bounds_2d[2] >= plane_bounds[2]
-                    and bounds_2d[3] >= plane_bounds[3]
+                    bbox[0, 0] <= bounds[0, 0]
+                    and bbox[0, 1] <= bounds[0, 1]
+                    and bbox[1, 0] >= bounds[1, 0]
+                    and bbox[1, 1] >= bounds[1, 1]
                 ):
-                    struct_contain_inds.append(ind)
-            struct_contains.append(struct_contain_inds)
-
-            # figure out where to place the bounding box coordinates of current structure
-            coord_min, coord_max = bmin[axis], bmax[axis]
-            indsmin = np.argwhere(coord_min < interval_coords)
-            indsmax = np.argwhere(coord_max > interval_coords)
-
-            # Exit if structure is outside of domain bounds
-            if (
-                indsmin.size == 0
-                or indsmax.size == 0
-                or np.any(bounds_2d[:2] >= sim_plane_bbox[2:])
-                or np.any(bounds_2d[2:] <= sim_plane_bbox[:2])
-            ):
-                continue
+                    struct_contains_inds.append(ind)
+            struct_contains.append(struct_contains_inds)
 
-            # Add current structure bounding box coordinates
+            # Figure out where to place the bounding box coordinates of current structure
+            indsmin = np.argwhere(bbox[0, 2] < interval_coords)
             indmin = int(indsmin[0])
-            indmax = int(indsmax[-1]) + 2
-            interval_coords = np.insert(interval_coords, indmin, coord_min)
-            interval_coords = np.insert(interval_coords, indmax, coord_max)
-            # Copy the structure containment list to the newly created interval
-            structs_list_copy = interval_structs[max(0, indmin - 1)].copy()
-            interval_structs.insert(indmin, structs_list_copy)
-            structs_list_copy = interval_structs[min(indmax - 1, len(interval_structs) - 1)].copy()
-            interval_structs.insert(indmax, structs_list_copy)
+            bbox0 = np.copy(bbox)
+            bbox0[1, 2] = bbox0[0, 2]
+            if not self.is_contained(bbox0, struct_bbox[struct_ind + 1 :]):
+                # Add current structure bounding box coordinates
+                interval_coords = np.insert(interval_coords, indmin, bbox[0, 2])
+                # Copy the structure containment list to the newly created interval
+                struct_list = interval_structs[max(0, indmin - 1)]
+                interval_structs.insert(indmin, struct_list.copy())
+
+            indsmax = np.argwhere(bbox[1, 2] > interval_coords)
+            indmax = int(indsmax[-1])
```

Comment:
One possible issue here is the possibility of `indsmax=[]`, which can arise e.g. when looking at mesh at x=0 but there is a structure of zero size along x and placed at x=0. I added the following lines to make it work, but not sure if it's correct,
  ```
 if len(indsmax)==0:
                continue
```

Reply:
I think this cannot happen, because previously we have made sure that only structures that are at least partially in the simulation domain are being meshed. Therefore, the "right" edge of a given structure has to be at least larger than the left edge of the simulation domain, which is the first element in `interval_coords`.